File: 2022/13/solution.pl
#!perl
use v5.16;
use warnings;

sub my_cmp {
	my ( $x, $y ) = @_;
	return $y <=> $x          if ( !ref $x and !ref $y );
	return my_cmp( [$x], $y ) if ( !ref $x and  ref $y );
	return my_cmp( $x, [$y] ) if (  ref $x and !ref $y );
	my $i = 0;
	while () {
		return  0 if ( $i > $#$x and $i > $#$y );
		return  1 if ( $i > $#$x );
		return -1 if ( $i > $#$y );
		my $cmp = my_cmp( $x->[$i], $y->[$i] );
		return $cmp if $cmp;
		++$i;
	}
}

my $filename = 'input.txt';

PART_1: {
	my ( $i, $total ) = ( 0, 0 );
	open( my $fh, '<', $filename );
	while ( ++$i and not eof $fh ) {
		my $first  = eval( scalar <$fh> );
		my $second = eval( scalar <$fh> );
		my $blank  = scalar <$fh>;
		$total += $i unless my_cmp( $first, $second ) < 0;
	}
	say "Index total: $total";
}

PART_2: {
	open( my $fh, '<', $filename );
	my @all = map { chomp; length($_) ? eval($_) : () } <$fh>;
	my @markers = ( [[2]], [[6]] );
	push @all, @markers;
	my @sorted = sort { my_cmp($b, $a) } @all;
	my @indices = grep {
		$sorted[$_-1]==$markers[0] or $sorted[$_-1]==$markers[1]
	} 1 .. @sorted;
	say "Decoder key: ", $indices[0] * $indices[1];
}

File: 2022/13/solution.php
<?php

function my_cmp ( $x, $y ) {
	if ( is_int($x) and is_int($y) )
		return $y <=> $x;
	if ( is_int($x) and is_array($y) )
		return my_cmp( [$x], $y );
	if ( is_array($x) and is_int($y) )
		return my_cmp( $x, [$y] );

	$i = 0;
	while ( true ) {
		if ( $i >= count($x) and $i >= count($y) )
			return 0;
		if ( $i >= count($x) )
			return 1;
		if ( $i >= count($y) )
			return -1;
		$cmp = my_cmp( $x[$i], $y[$i] );
		if ( $cmp != 0 )
			return $cmp;
		++$i;
	}
}

$filename = 'input.txt';

PART_1: {
	list ( $i, $total ) = [ 0, 0 ];
	$lines = file( $filename );
	while ( count($lines) ) {
		$first  = json_decode( array_shift($lines) );
		$second = json_decode( array_shift($lines) );
		array_shift($lines);
		++$i;
		if ( my_cmp( $first, $second ) >= 0 )
			$total += $i;
	}
	echo "Index total: $total\n";
}

PART_2: {
	$all = array_map(
		function ( $line ) { return json_decode( $line ); },
		array_filter(
			file( $filename ),
			function ( $line ) { return preg_match( '/\S/', $line ); },
		),
	);
	$all []= [[2]];
	$all []= [[6]];
	usort( $all, 'my_cmp' );
	$all = array_reverse( $all );
	$indices = [];
	foreach ( $all as $ix => $line ) {
		$json = json_encode( $line );
		if ( $json === '[[2]]' or $json === '[[6]]' ) $indices []= 1 + $ix;
	}
	echo "Decoder key: " . ( $indices[0] * $indices[1] ) . "\n";
}
